neurots/generate/orientations.py

```diff
@@ -360,3 +435,215 @@ def compute_interval_n_tree(soma, n_trees, rng=np.random):
         interval_n_trees = np.array([n_trees])
 
     return phi_intervals, interval_n_trees
+
+
+def spherical_angles_to_pia_orientations(phis, thetas):
+    """Compute orientation from spherical angles where thetas are wrt to pia at `[0, 1, 0]`.
+
+    Args:
+        phis (numpy.ndarray): Polar angles.
+        thetas (numpy.ndarray): Azimuthal angles.
+
+    Returns:
+        numpy.ndarray: The orientation vectors where each row corresponds to a phi-theta pair.
+    """
+    return np.column_stack(
+        (np.cos(phis) * np.sin(thetas), np.cos(thetas), np.sin(phis) * np.sin(thetas))
+    )
+
+
+def get_probability_function(form="step", with_density=True):
+    """Get probability functions to fit 3d trunk angles distributions.
+
+    Args:
+        form (str): Form of the function, can be `flat`, `step` or `double_step`.
+        with_density (bool): Return the function with spherical density factor or not.
+
+    Three forms of functions are available:
+    - `flat`: uniform flat distribution
+    - `step`: distribution with a single sigmoid :func:`scipy.special.expit`
+    - `double_step`: distribution with two opposite sigmoids :func:`scipy.special.expit`
+
+    Each sigmoid is parametrized by a scale and a rate.
+
+    In practice, the `flat` function is used when no asymetry is present in the data, and the other
+    two are when an asymmetry towards one direction, usually opposite to pia or apical,
+    or two directions, usually along and opposite to pia.
+
+    Returns:
+        function with first arg as angle and next args to parametrize the function
+    """
+    if form == "flat":
+
+        def flat_prob(angle):
+            p = 1.0 + 0 * angle
+            if with_density:
+                p *= np.sin(angle)
+            return p
+
+        return flat_prob
+
+    if form == "step":
+
+        def single_prob(angle, scale, rate):
+            def _prob(angle):
+                return expit((angle - scale) / rate)
+
+            p = _prob(angle) / max(_prob(np.linspace(0, np.pi, 100)))
+            if with_density:
+                p *= np.sin(angle)
+            return np.clip(p, 0.0, 1.0)
+
+        return single_prob
+
+    if form == "double_step":
+
+        def double_prob(angle, scale_low, rate_low, scale_high, rate_high):
+            def _prob(angle):
+                return 0.5 * (
+                    expit((angle - scale_low) / rate_low) + expit((-angle - scale_high) / rate_high)
+                )
+
+            p = _prob(angle) / max(_prob(np.linspace(0, np.pi, 100)))
+            if with_density:
+                p *= np.sin(angle)
+            return np.clip(p, 0.0, 1.0)
+
+        return double_prob
+
+    raise ValueError(
+        f"The '{form}' value is unknown, it should be one of ['flat', 'step', 'double_step']"
+    )
+
+
+def _fit_single_3d_angles(data, neurite_type, morph_class, fit_params=None):
```

Comment:
morph_class is not very comprehensive, maybe apical_present instead or if this can be "True" "False" you can use has_apical

Reply:
sure, but this name is from taxonomy file, which l looks like:
```
mtype   mClass  sClass
L1_DAC  INT INH
L1_HAC  INT INH
```